File: solutions/ruby/bob/1/bob.rb
module Bob
    RESPONSE = {
        whatever: 'Whatever.',
        chill: 'Whoa, chill out!',
        sure: 'Sure.',
        calm: "Calm down, I know what I'm doing!",
        fine: 'Fine. Be that way!'
    }.freeze

    def self.hey(message)
        message = message.strip
        return RESPONSE[:fine] if message.empty?
        return RESPONSE[:calm] if question?(message) && shouting?(message)
        return RESPONSE[:sure] if question?(message)
        return RESPONSE[:chill] if shouting?(message)
        RESPONSE[:whatever]
    end

    private
    def self.question?(message)
        message.end_with?('?')
    end

    def self.shouting?(message)
        cleaned = message.gsub(/[^a-z]/i, '')
        !cleaned.empty? && cleaned.upcase == cleaned
    end
end

File: solutions/python/bob/1/bob.py
RESPONSE = {
    'whatever': 'Whatever.',
    'chill': 'Whoa, chill out!',
    'sure': 'Sure.',
    'calm': "Calm down, I know what I'm doing!",
    'fine': 'Fine. Be that way!'
}

def is_question(message: str) -> bool:
    return message.strip().endswith('?')

def is_shouting(message: str) -> bool:
    cleaned = ''.join(char for char in message if char.isalpha())
    return cleaned.isupper() and cleaned

def response(message: str) -> str:
    message = message.strip()
    if not message:
        return RESPONSE['fine']
    if is_question(message) and is_shouting(message):
        return RESPONSE['calm']
    if is_question(message):
        return RESPONSE['sure']
    if is_shouting(message):
        return RESPONSE['chill']
    return RESPONSE['whatever']